paramak/parametric_shapes/extruded_mixed_shape.py

```diff
@@ -0,0 +1,92 @@
+
+from paramak import Shape
+from paramak.utils import calculate_wedge_cut
+
+
+class ExtrudeMixedShape(Shape):
+    """Extrudes a 3d CadQuery solid from points connected with a mixture of
+    straight and spline connections.
+
+    Args:
+        distance (float): the extrusion distance to use (cm units if used for
+            neutronics)
+        extrude_both (bool, optional): If set to True, the extrusion will
+            occur in both directions. Defaults to True.
+        rotation_angle (float, optional): rotation angle of solid created. A
+            cut is performed from rotation_angle to 360 degrees. Defaults to
+            360.0.
+        stp_filename (str, optional): Defaults to "ExtrudeMixedShape.stp".
+        stl_filename (str, optional): Defaults to "ExtrudeMixedShape.stl".
+
+    """
+
+    def __init__(
+        self,
+        distance,
+        extrude_both=True,
+        rotation_angle=360.0,
+        extrusion_start_offset=0.0,
+        stp_filename="ExtrudeMixedShape.stp",
+        stl_filename="ExtrudeMixedShape.stl",
+        **kwargs
+    ):
+
+        super().__init__(
+            stp_filename=stp_filename,
+            stl_filename=stl_filename,
+            **kwargs
+        )
+        self.distance = distance
+        self.extrude_both = extrude_both
+        self.rotation_angle = rotation_angle
+        self.extrusion_start_offset = extrusion_start_offset
+
+    @property
+    def distance(self):
+        return self._distance
+
+    @distance.setter
+    def distance(self, value):
+        self._distance = value
+
+    @property
+    def rotation_angle(self):
+        return self._rotation_angle
+
+    @rotation_angle.setter
+    def rotation_angle(self, value):
+        self._rotation_angle = value
+
+    @property
+    def extrusion_start_offset(self):
+        return self._extrusion_start_offset
+
+    @extrusion_start_offset.setter
+    def extrusion_start_offset(self, value):
+        self._extrusion_start_offset = value
+
+    def create_solid(self):
+        """Creates an extruded 3d solid using points connected with straight
+        and spline edges.
+
+           Returns:
+              A CadQuery solid: A 3D solid volume
+        """
+
+        solid = super().create_solid()
+
+        if not self.extrude_both:
+            extrusion_distance = -self.distance
+        else:
+            extrusion_distance = -self.distance / 2.0
+
+        solid = solid.close().extrude(
+            distance=extrusion_distance,
+            both=self.extrude_both)
+
+        solid = self.rotate_solid(solid)
+        cutting_wedge = calculate_wedge_cut(self)
```

Comment:
**BEST PRACTICE**

The `ExtrudeMixedShape` class calls from `paramak.utils` import `calculate_wedge_cut` here.

This util function is defined in `utils.py` line 20.

The util function itself takes a `self` argument which relies on a `rotation_angle` attribute. Normally, the way to do this in object-oriented programming is to define a base class (or mixin) that has this `rotation_angle` attribute but also defines the `calculate_wedge_cut` function so that you don't have to import it from another module. You can simply call it from `self` like:

```
        solid = self.rotate_solid(solid)
        cutting_wedge = self.calculate_wedge_cut() # Call it as an instance method rather than as a global function
        solid = self.perform_boolean_operations(solid, wedge_cut=cutting_wedge)
        self.solid = solid
```
<p align="right">
    <a href="https://www.pullrequest.com/">
        <img width="15%" src="https://static.pullrequest.com/images/dashboard/pullrequest-logo.svg">
    </a>
</p>